packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -2176,90 +2175,25 @@ function dispatchReducerAction<S, A>(
-    didScheduleRenderPhaseUpdateDuringThisPass = didScheduleRenderPhaseUpdate = true;
-    const pending = queue.pending;
-    if (pending === null) {
-      // This is the first update. Create a circular list.
-      update.next = update;
-    } else {
-      update.next = pending.next;
-      pending.next = update;
-    }
-    queue.pending = update;
+  if (isRenderPhaseUpdate(fiber)) {
+    enqueueRenderPhaseUpdate(queue, update);
   } else {
-    if (isInterleavedUpdate(fiber, lane)) {
-      const interleaved = queue.interleaved;
-      if (interleaved === null) {
-        // This is the first update. Create a circular list.
-        update.next = update;
-        // At the end of the current render, this queue's interleaved updates will
-        // be transferred to the pending queue.
-        pushInterleavedQueue(queue);
-      } else {
-        update.next = interleaved.next;
-        interleaved.next = update;
-      }
-      queue.interleaved = update;
-    } else {
-      const pending = queue.pending;
-      if (pending === null) {
-        // This is the first update. Create a circular list.
-        update.next = update;
-      } else {
-        update.next = pending.next;
-        pending.next = update;
-      }
-      queue.pending = update;
-    }
+    enqueueUpdate(fiber, queue, update, lane);
 
     if (__DEV__) {
       // $FlowExpectedError - jest isn't a global, and isn't recognized outside of tests
       if ('undefined' !== typeof jest) {
         warnIfNotCurrentlyActingUpdatesInDev(fiber);
       }
     }
+    const eventTime = requestEventTime();
     const root = scheduleUpdateOnFiber(fiber, lane, eventTime);
-
-    if (isTransitionLane(lane) && root !== null) {
-      let queueLanes = queue.lanes;
-
-      // If any entangled lanes are no longer pending on the root, then they
-      // must have finished. We can remove them from the shared queue, which
-      // represents a superset of the actually pending lanes. In some cases we
-      // may entangle more than we need to, but that's OK. In fact it's worse if
-      // we *don't* entangle when we should.
-      queueLanes = intersectLanes(queueLanes, root.pendingLanes);
-
-      // Entangle the new transition lane with the other transition lanes.
-      const newQueueLanes = mergeLanes(queueLanes, lane);
-      queue.lanes = newQueueLanes;
-      // Even if queue.lanes already include lane, we don't know for certain if
-      // the lane finished since the last time we entangled it. So we need to
-      // entangle it again, just to be sure.
-      markRootEntangled(root, newQueueLanes);
-    }
-  }
-
-  if (__DEV__) {
-    if (enableDebugTracing) {
-      if (fiber.mode & DebugTracingMode) {
-        const name = getComponentNameFromFiber(fiber) || 'Unknown';
-        logStateUpdateScheduled(name, lane, action);
-      }
+    if (root !== null) {
+      entangleTransitionUpdate(root, queue, lane);
```

Comment:
One suggestion here: since `entangleTransitionUpdate()` immediately checks `isTransitionLane(lane)`, it might be better to put the check outside and avoid calling the function unless `isTransitionLane(lane)` is true. Something like:

```
// Takes a lane, checks if its entangled, returns the lane as an opaque type wrapper if so
const entangledLane: EntangledLane | null = checkEntangledLane(lane);
if (entangedLand !== null) {
  // type of entangledLane is 
  // opaque type EntangledLane: Lane = Lane;
  entangleTransitionUpdate(root, queue, entangledLane);
}
```

This doesn't incur any overhead - entangedLane is still a number - but it uses types to enforce correct calling sequence. This also lets the JIT inline just `checkEntangledLane()`, and potentially avoid inlining the more rarely called `entangleTransitionUpdate()`.

Maybe not worth it but figured i'd mention it.

Reply:
Lane used to be an opaque type but I had to un-opaque it because it leaks outside of the reconciler via the Fiber type, unfortunately. And the Fiber type is shared between reconcilers. Our Flow types for Fiber are kind of a mess, not even sure if it's possible to model correctly and if it were it would involve so much work that we might as well just go all the way to Rust.

Re: avoiding the function call, Closure inlines both of these functions anyway, but that's a good point. I had it that way at first but I moved it for for symmetry with the equivalent class update queue function. I figure I'll hoist it out as soon as we need to put additional transition-only logic in there.